src/ShellWords.hs

```diff
@@ -28,50 +26,34 @@ parse :: String -> Either String [String]
 parse = runParser parser
 
 runParser :: Parser a -> String -> Either String a
-runParser p = first errorBundlePretty . Megaparsec.parse p "<input>"
+runParser p = first errorBundlePretty . Megaparsec.parse (p <* eof) "<input>"
 
 -- | Parse and return @'Text'@ values
 parseText :: Text -> Either String [Text]
 parseText = fmap (map pack) . parse . unpack
 
 parser :: Parser [String]
-parser = fmap (dropWhileEnd null) $ space *> shellword `sepEndBy1` space1
+parser = space *> shellword `sepEndBy1` space1 <* space
 
 shellword :: Parser String
-shellword =
-  choice
-    [ quoted <?> "quoted string"
-    , shelloption <?> "shell option"
-    , value <?> "bare value"
-    ]
-    <?> "shell word"
+shellword = fmap concat $ some $ bare <|> quoted
```

Comment:
Very neat that handling string-adjacency like this actually simplified everything.